dbms/src/Common/FiberRWLock.h

```diff
@@ -0,0 +1,101 @@
+// Copyright 2022 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <boost/fiber/all.hpp>
+
+namespace DB
+{
+/// FiberRWLock has the same interface with `std::shared_mutex` and is intended to
+/// work with `std::shared_lock` or `std::unique_lock`.
+/// FiberRWLock is designed to work well with both OS thread and boost::fiber, though the
+/// latter performs better.
+///
+/// It is designed as writer-first in that waiting writer will block following readers.
+/// TODO: if there's a new coming writer after some blocking readers, current implementation
+/// will let this writer go before readers. This simplifies the code but hurts the fairness.
+class FiberRWLock
+{
+public:
+    FiberRWLock() = default;
+    FiberRWLock(const FiberRWLock &) = delete;
+    ~FiberRWLock() = default;
+
+    FiberRWLock & operator=(const FiberRWLock &) = delete;
+
+    void lock()
+    {
+        std::unique_lock lock(mu);
+        ++writer_wait_cnt;
+        writer_cv.wait(lock, [&] { return writer_cnt == 0 && reader_cnt == 0; });
```

Comment:
I worry about the fairness of readers in this implementation.
For instance：
1. a writer holds a lock and some readers wait for this lock
2. another writer comes and adds `writer_wait_cnt`.
3. the first writer releases the lock and the second writer will get the lock no matter how long these readers wait.

I saw another implementation of rwlock in cmudb(cmu 15-445) that solves this problem.https://paste.ofcode.org/jKfGZ23H977wszmgmPMRdW
The writers and readers wait on `reader condition variable` first. So in above case, the second writer can not get the lock because it is the latter one.

Reply:
Another reference, the implementation of `shared_mutex` https://code.woboq.org/llvm/libcxx/src/shared_mutex.cpp.html, it is basically the same as https://paste.ofcode.org/jKfGZ23H977wszmgmPMRdW.